<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_161) on Fri Aug 30 15:33:45 EDT 2019 -->
<title>CMAESSpecies</title>
<meta name="date" content="2019-08-30">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="CMAESSpecies";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../ec/eda/cmaes/CMAESInitializer.html" title="class in ec.eda.cmaes"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?ec/eda/cmaes/CMAESSpecies.html" target="_top">Frames</a></li>
<li><a href="CMAESSpecies.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">ec.eda.cmaes</div>
<h2 title="Class CMAESSpecies" class="title">Class CMAESSpecies</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li><a href="../../../ec/Species.html" title="class in ec">ec.Species</a></li>
<li>
<ul class="inheritance">
<li><a href="../../../ec/vector/VectorSpecies.html" title="class in ec.vector">ec.vector.VectorSpecies</a></li>
<li>
<ul class="inheritance">
<li><a href="../../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">ec.vector.FloatVectorSpecies</a></li>
<li>
<ul class="inheritance">
<li>ec.eda.cmaes.CMAESSpecies</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="../../../ec/Prototype.html" title="interface in ec">Prototype</a>, <a href="../../../ec/Setup.html" title="interface in ec">Setup</a>, java.io.Serializable, java.lang.Cloneable</dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">CMAESSpecies</span>
extends <a href="../../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></pre>
<div class="block">CMAESSpecies is a FloatVectorSpecies which implements a faithful version of the
 CMA-ES algorithm.  The class has two basic methods.  The newIndividual(...)
 method generates a new random individual underneath the current CMA-ES
 gaussian distribution.  The updateDistribution(...) method revises the
 gaussian distribution to reflect the fitness results of the population.
 
 <p>CMAESSpecies must be used in combination with CMAESBreeder, which will
 call it at appropriate times to revise the distribution and to generate a new
 subpopulation of individuals.  It must also be used in combination with
 CMAESInitializer, which will use it to generate the initial population.

 <p>Note importantly that CMAESSpecies <b>ignores the subpopulation size</b>.
 Instead the first thing it will do is revise the subpopulation size to reflect
 the "lambda" parameter.  This is a consequence of another feature of
 CMAESSpecies: many of its parameters do not have fixed default values, but
 rather values which are computed on the fly if the user does not provide them.
 For this reason, it also prints out these values when running so the user may
 see what values it used for that given run.  The computed default values 
 use equations which are common in the CMA-ES literature and are described
 below.

 <p>CMAESSpecies also has an "alternative termination" option.  Normally ECJ
 terminates when the optimal individual is discovered or when the generations
 or maximum number of evaluations has been exceeded.  CMA-ES can optionally
 terminate when the eigenvalues of the covariance matrix of its gaussian
 distribution are too small.  Among other things, this will generally prevent
 CMA-ES from terminating abnormally because an eigenvalue has gone negative
 (due to floating point underflows).  But by default this alternative termination
 is turned off, and CMA-ES will simply terminate in that case.

 <p>CMAESSpecies needs initial values of sigma (the scaling parameter for its
 covariance matrix) and the mean of the gaussian distribution.  By default
 sigma's initial value is 1.0.  The mean must be set to one of "zero"
 (meaning the origin), "center" (meaning the center of the genome space defined
 by the min and max gene values for each gene), or "random" (meaning a randomly-
 chosen point in the space".  If it is not set to any of these, you may
 alternatively set the initial mean values by hand.  But you must do one of
 the two.

 <p>Initializing the covariance matrix can be a problem in in CMA-ES, particularly
 if it is large relative to the gene bounds.  If CMA-ES generates a random individual
 under its current distribution and that individual violates the bounds of just a 
 single gene, it is invalid and must be regenerated.  If you have a lot of genes,
 and the covariance matrix is large relative to their bounds, then the probability
 that this will occur rapidly approaches 1.0, so CMA-ES will be trapped in an effectively
 infinite loop endlessly producing invalid individuals.

 <p>This can be remedied in a few ways.  First there is an option available to force
 the initial covariance matrix to NOT be the identity matrix (the default) but instead
 be scaled according to the gene bounds.  That may help.  You can also of course reduce
 sigma.  Last, you can turn on an alternative individual generation mechanism; here,
 if a specific gene bound is violated, then *for that gene only* the value is chosen at
 random uniformly from within the gene bounds.

 <p>CMAESSpecies relies on the EJML matrix library, available at 
 <a href="http://ejml.org/">http://ejml.org/</a>

 <p><b>Parameters</b><br>
 <table>
 <tr><td valign=top><i>base</i>.<tt>mean</tt><br>
 <font size=-1>String, one of center, zero, or random</font></td>
 <td valign=top>(the initial mean for the distribution)<br>
 If not provided, <i>base</i>.<tt>mean</tt>.0 and so on must be given    
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>mean</tt><br>
 <font size=-1>String, one of center, zero, or random</font></td>
 <td valign=top>(the initial mean for the distribution)<br>
 If not provided, <i>base</i>.<tt>mean</tt>.0 and so on must be given    
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>mean</tt>.<i>i</i><br>
 <font size=-1>Floating-point value</font></td>
 <td valign=top>(the value of dimension i of the initial mean vector)<br>
 These values will override those set as a result of <i>base</i>.<tt>mean</tt> 
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>sigma</tt><br>
 <font size=-1>Floating-point value > 0.0</font></td>
 <td valign=top>(the sigma scaling parameter)<br>
 If not provided, this defaults to 1.0.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>lambda</tt><br>
 <font size=-1>Integer > <i>base</i>.<tt>mu</tt></font></td>
 <td valign=top>(lambda population size)<br>
 If not provided, this defaults to 4 + Math.floor(3 * Math.log(n)).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>lambda</tt><br>
 <font size=-1>Integer > 1</td>
 <td valign=top>(mu truncated population size)<br>
 If not provided, this defaults to Math.floor(lambda / 2.0).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>weight</tt>.<i>i</i><br>
 <font size=-1>Float >= 0</td>
 <td valign=top>(weight for individual i (i from 0 to mu-1))<br>
 If not provided, this defaults to Math.log((lambda + 1.0) / (2.0 * (i + 1))).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>cc</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(c_c parameter)<br>
 If not provided, this defaults to (4.0+mueff/n) / (n+4.0 + 2.0*mueff/n)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>
 
 <tr><td valign=top><i>base</i>.<tt>cs</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(c_sigma parameter)<br>
 If not provided, this defaults to (mueff+2.0)/(n+mueff+5.0)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>c1</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1 (and c1 > (1-cmu))</td>
 <td valign=top>(c_1 parameter)<br>
 If not provided, this defaults to 2.0 / ((n+1.3)*(n+1.3)+mueff)<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>cmu</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1 (and cmu > (1-c1))</td>
 <td valign=top>(c_mu parameter)<br>
 If not provided, this defaults to Math.min(1.0-c1, 2.0*(mueff-2.0+1.0/mueff) / ((n+2.0)*(n+2.0)+mueff))<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>damps</tt><br>
 <font size=-1>0 &lt;= Float &lt; 1</td>
 <td valign=top>(d_sigma dampening parameter)<br>
 If not provided, this defaults to 1.0 + 2.0*Math.max(0.0, Math.sqrt((mueff-1.0)/(n+1.0))-1.0) + cs<br>
 Where mueff is defined in the variables below, and n is the genome size
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>covariance</tt><br>
 <font size=-1>String, either "identity" (default) or "scaled"</td>
 <td valign=top>Covariance matrix initialization procedure.<br>
 If "identity", then the covariance matrix is initialized to the
 identity matrix.  If "scaled", then the covariance matrix is
 initialized to a diagonal matrix whose values are squares of
 each gene range (max - min).
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-generator</tt><br>
 <font size=-1>true or false (default)</td>
 <td valign=top>Whether or not to use the alternative indivdiual-generation procedure.<br>
 If "true", then if, in the process of generating an individual, we have failed 
 alternative-generator-tries times to create an individual which falls within the min and
 max gene values for each gene, then whenever a gene value violates those constraints
 we will simply randomize it to something uniformly chosen between the min and max.
 If "false", then whenever an individual violates constraints, we will try again as
 necessary.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-generator-tries</tt><br>
 <font size=-1>Integer > 1 (default is 100)</td>
 <td valign=top>How many times we try to generate a valid individual before
 possibly using the alternative-generator approach.
 </td></tr>

 <tr><td valign=top><i>base</i>.<tt>alternative-termination</tt><br>
 <font size=-1>boolean, default false</td>
 <td valign=top>Should we also terminate on CMA-ES's additional termination conditions?.
 </td></tr>

 </table>


 <p><b>Default Base</b><br>
 eda.cma-es.species</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../serialized-form.html#ec.eda.cmaes.CMAESSpecies">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#altGeneratorTries">altGeneratorTries</a></span></code>
<div class="block">How many times should we try to generate a valid individual before we give up and use the useAltGenerator approach?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#b">b</a></span></code>
<div class="block">The "B" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>org.ejml.data.DenseMatrix64F</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#bd">bd</a></span></code>
<div class="block">b x d</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#c">c</a></span></code>
<div class="block">The "C" covariance matrix of the distribution.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#c1">c1</a></span></code>
<div class="block">The c_1 rank-1 update learning rate.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#cc">cc</a></span></code>
<div class="block">The c_c rank-one evolution path cumulation parameter.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#chiN">chiN</a></span></code>
<div class="block">An estimate of the expected size of the standard multivariate gaussian N(0,I).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#cmu">cmu</a></span></code>
<div class="block">The c_{\mu} rank-mu update learning rate.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#cs">cs</a></span></code>
<div class="block">The c_{\sigma} mutation rate evolution path learning rate.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#d">d</a></span></code>
<div class="block">The "C" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#damps">damps</a></span></code>
<div class="block">The d_{\sigma} dampening for the mutation rate update.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#DEFAULT_ALT_GENERATOR_TRIES">DEFAULT_ALT_GENERATOR_TRIES</a></span></code>
<div class="block">Default value (100) for altGeneratorTries.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#invsqrtC">invsqrtC</a></span></code>
<div class="block">C^{-1/2}.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#lambda">lambda</a></span></code>
<div class="block">The individuals generated from the distribution.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#lastEigenDecompositionGeneration">lastEigenDecompositionGeneration</a></span></code>
<div class="block">The most recent generation where an eigendecomposition on C was performed into B and D</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#MAX_TRIES_BEFORE_WARNING">MAX_TRIES_BEFORE_WARNING</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#mu">mu</a></span></code>
<div class="block">The truncated individuals used to update the distribution.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#mueff">mueff</a></span></code>
<div class="block">The "mu_{eff}" constant in CMA-ES.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_ALTERNATIVE_GENERATOR">P_ALTERNATIVE_GENERATOR</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_ALTERNATIVE_GENERATOR_TRIES">P_ALTERNATIVE_GENERATOR_TRIES</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_ALTERNATIVE_TERMINATION">P_ALTERNATIVE_TERMINATION</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_C1">P_C1</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_CC">P_CC</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_CMAES_SPECIES">P_CMAES_SPECIES</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_CMU">P_CMU</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_COVARIANCE">P_COVARIANCE</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_CS">P_CS</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_DAMPS">P_DAMPS</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_LAMBDA">P_LAMBDA</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_MEAN">P_MEAN</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_MU">P_MU</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_SIGMA">P_SIGMA</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#P_WEIGHTS">P_WEIGHTS</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#pc">pc</a></span></code>
<div class="block">The p_c evolution path vector.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#ps">ps</a></span></code>
<div class="block">The p_{\sigma} evolution path vector.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.data.DenseMatrix64F</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#sbd">sbd</a></span></code>
<div class="block">bd x sigma</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#sigma">sigma</a></span></code>
<div class="block">The "sigma" scaling factor for the covariance matrix.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#useAltGenerator">useAltGenerator</a></span></code>
<div class="block">If, after trying altGeneratorTries to build an indiviual, we are still building one which violates min/max gene constraints, should
        we instead fill those violated genes with uniformly-selected values between the min and max?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#useAltTermination">useAltTermination</a></span></code>
<div class="block">Should we terminate when the eigenvalues get too small?  If we don't, they might go negative and the eigendecomposition will fail.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#V_CENTER">V_CENTER</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#V_IDENTITY">V_IDENTITY</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#V_RANDOM">V_RANDOM</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#V_SCALED">V_SCALED</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#V_ZERO">V_ZERO</a></span></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#weights">weights</a></span></code>
<div class="block">The ranked fitness weights for the mu individuals.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>org.ejml.simple.SimpleMatrix</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#xmean">xmean</a></span></code>
<div class="block">The mean of the distribution.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.ec.vector.FloatVectorSpecies">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.vector.<a href="../../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></h3>
<code><a href="../../../ec/vector/FloatVectorSpecies.html#C_GAUSS_MUTATION">C_GAUSS_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#C_INTEGER_RANDOM_WALK_MUTATION">C_INTEGER_RANDOM_WALK_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#C_INTEGER_RESET_MUTATION">C_INTEGER_RESET_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#C_POLYNOMIAL_MUTATION">C_POLYNOMIAL_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#C_RESET_MUTATION">C_RESET_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#DEFAULT_OUT_OF_BOUNDS_RETRIES">DEFAULT_OUT_OF_BOUNDS_RETRIES</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#gaussMutationStdev">gaussMutationStdev</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#maxGene">maxGene</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#minGene">minGene</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationDistributionIndex">mutationDistributionIndex</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationIsBounded">mutationIsBounded</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationType">mutationType</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#outOfBoundsRetries">outOfBoundsRetries</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_MAXGENE">P_MAXGENE</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_MINGENE">P_MINGENE</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_MUTATION_BOUNDED">P_MUTATION_BOUNDED</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_MUTATION_DISTRIBUTION_INDEX">P_MUTATION_DISTRIBUTION_INDEX</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_MUTATIONTYPE">P_MUTATIONTYPE</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_OUTOFBOUNDS_RETRIES">P_OUTOFBOUNDS_RETRIES</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_POLYNOMIAL_ALTERNATIVE">P_POLYNOMIAL_ALTERNATIVE</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_RANDOM_WALK_PROBABILITY">P_RANDOM_WALK_PROBABILITY</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#P_STDEV">P_STDEV</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#polynomialIsAlternative">polynomialIsAlternative</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#randomWalkProbability">randomWalkProbability</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#V_GAUSS_MUTATION">V_GAUSS_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#V_INTEGER_RANDOM_WALK_MUTATION">V_INTEGER_RANDOM_WALK_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#V_INTEGER_RESET_MUTATION">V_INTEGER_RESET_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#V_POLYNOMIAL_MUTATION">V_POLYNOMIAL_MUTATION</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#V_RESET_MUTATION">V_RESET_MUTATION</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.ec.vector.VectorSpecies">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.vector.<a href="../../../ec/vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../../ec/vector/VectorSpecies.html#C_ANY_POINT">C_ANY_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#C_GEOMETRIC">C_GEOMETRIC</a>, <a href="../../../ec/vector/VectorSpecies.html#C_INTERMED_RECOMB">C_INTERMED_RECOMB</a>, <a href="../../../ec/vector/VectorSpecies.html#C_LINE_RECOMB">C_LINE_RECOMB</a>, <a href="../../../ec/vector/VectorSpecies.html#C_NONE">C_NONE</a>, <a href="../../../ec/vector/VectorSpecies.html#C_ONE_POINT">C_ONE_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#C_ONE_POINT_NO_NOP">C_ONE_POINT_NO_NOP</a>, <a href="../../../ec/vector/VectorSpecies.html#C_SIMULATED_BINARY">C_SIMULATED_BINARY</a>, <a href="../../../ec/vector/VectorSpecies.html#C_TWO_POINT">C_TWO_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#C_TWO_POINT_NO_NOP">C_TWO_POINT_NO_NOP</a>, <a href="../../../ec/vector/VectorSpecies.html#C_UNIFORM">C_UNIFORM</a>, <a href="../../../ec/vector/VectorSpecies.html#chunksize">chunksize</a>, <a href="../../../ec/vector/VectorSpecies.html#crossoverDistributionIndex">crossoverDistributionIndex</a>, <a href="../../../ec/vector/VectorSpecies.html#crossoverProbability">crossoverProbability</a>, <a href="../../../ec/vector/VectorSpecies.html#crossoverType">crossoverType</a>, <a href="../../../ec/vector/VectorSpecies.html#duplicateRetries">duplicateRetries</a>, <a href="../../../ec/vector/VectorSpecies.html#dynamicInitialSize">dynamicInitialSize</a>, <a href="../../../ec/vector/VectorSpecies.html#genomeIncreaseProbability">genomeIncreaseProbability</a>, <a href="../../../ec/vector/VectorSpecies.html#genomeResizeAlgorithm">genomeResizeAlgorithm</a>, <a href="../../../ec/vector/VectorSpecies.html#genomeSize">genomeSize</a>, <a href="../../../ec/vector/VectorSpecies.html#lineDistance">lineDistance</a>, <a href="../../../ec/vector/VectorSpecies.html#maxInitialSize">maxInitialSize</a>, <a href="../../../ec/vector/VectorSpecies.html#minInitialSize">minInitialSize</a>, <a href="../../../ec/vector/VectorSpecies.html#mutationProbability">mutationProbability</a>, <a href="../../../ec/vector/VectorSpecies.html#P_CHUNKSIZE">P_CHUNKSIZE</a>, <a href="../../../ec/vector/VectorSpecies.html#P_CROSSOVER_DISTRIBUTION_INDEX">P_CROSSOVER_DISTRIBUTION_INDEX</a>, <a href="../../../ec/vector/VectorSpecies.html#P_CROSSOVERPROB">P_CROSSOVERPROB</a>, <a href="../../../ec/vector/VectorSpecies.html#P_CROSSOVERTYPE">P_CROSSOVERTYPE</a>, <a href="../../../ec/vector/VectorSpecies.html#P_DUPLICATE_RETRIES">P_DUPLICATE_RETRIES</a>, <a href="../../../ec/vector/VectorSpecies.html#P_GENOMESIZE">P_GENOMESIZE</a>, <a href="../../../ec/vector/VectorSpecies.html#P_GEOMETRIC_PROBABILITY">P_GEOMETRIC_PROBABILITY</a>, <a href="../../../ec/vector/VectorSpecies.html#P_LINEDISTANCE">P_LINEDISTANCE</a>, <a href="../../../ec/vector/VectorSpecies.html#P_MUTATIONPROB">P_MUTATIONPROB</a>, <a href="../../../ec/vector/VectorSpecies.html#P_NUM_SEGMENTS">P_NUM_SEGMENTS</a>, <a href="../../../ec/vector/VectorSpecies.html#P_SEGMENT">P_SEGMENT</a>, <a href="../../../ec/vector/VectorSpecies.html#P_SEGMENT_END">P_SEGMENT_END</a>, <a href="../../../ec/vector/VectorSpecies.html#P_SEGMENT_START">P_SEGMENT_START</a>, <a href="../../../ec/vector/VectorSpecies.html#P_SEGMENT_TYPE">P_SEGMENT_TYPE</a>, <a href="../../../ec/vector/VectorSpecies.html#P_UNIFORM_MAX">P_UNIFORM_MAX</a>, <a href="../../../ec/vector/VectorSpecies.html#P_UNIFORM_MIN">P_UNIFORM_MIN</a>, <a href="../../../ec/vector/VectorSpecies.html#P_VECTORSPECIES">P_VECTORSPECIES</a>, <a href="../../../ec/vector/VectorSpecies.html#V_ANY_POINT">V_ANY_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#V_GEOMETRIC">V_GEOMETRIC</a>, <a href="../../../ec/vector/VectorSpecies.html#V_INTERMED_RECOMB">V_INTERMED_RECOMB</a>, <a href="../../../ec/vector/VectorSpecies.html#V_LINE_RECOMB">V_LINE_RECOMB</a>, <a href="../../../ec/vector/VectorSpecies.html#V_ONE_POINT">V_ONE_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#V_ONE_POINT_NO_NOP">V_ONE_POINT_NO_NOP</a>, <a href="../../../ec/vector/VectorSpecies.html#V_SIMULATED_BINARY">V_SIMULATED_BINARY</a>, <a href="../../../ec/vector/VectorSpecies.html#V_TWO_POINT">V_TWO_POINT</a>, <a href="../../../ec/vector/VectorSpecies.html#V_TWO_POINT_NO_NOP">V_TWO_POINT_NO_NOP</a>, <a href="../../../ec/vector/VectorSpecies.html#V_UNIFORM">V_UNIFORM</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.ec.Species">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.<a href="../../../ec/Species.html" title="class in ec">Species</a></h3>
<code><a href="../../../ec/Species.html#f_prototype">f_prototype</a>, <a href="../../../ec/Species.html#i_prototype">i_prototype</a>, <a href="../../../ec/Species.html#P_FITNESS">P_FITNESS</a>, <a href="../../../ec/Species.html#P_INDIVIDUAL">P_INDIVIDUAL</a>, <a href="../../../ec/Species.html#P_PIPE">P_PIPE</a>, <a href="../../../ec/Species.html#pipe_prototype">pipe_prototype</a></code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#CMAESSpecies--">CMAESSpecies</a></span>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#clone--">clone</a></span>()</code>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code><a href="../../../ec/util/Parameter.html" title="class in ec.util">Parameter</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#defaultBase--">defaultBase</a></span>()</code>
<div class="block">Returns the default base for this prototype.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code><a href="../../../ec/Individual.html" title="class in ec">Individual</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#newIndividual-ec.EvolutionState-int-">newIndividual</a></span>(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
             int&nbsp;thread)</code>
<div class="block">Provides a brand-new individual to fill in a population.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#setup-ec.EvolutionState-ec.util.Parameter-">setup</a></span>(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
     <a href="../../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../ec/eda/cmaes/CMAESSpecies.html#updateDistribution-ec.EvolutionState-ec.Subpopulation-">updateDistribution</a></span>(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                  <a href="../../../ec/Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</code>
<div class="block">Revises the CMA-ES distribution to reflect the current fitness results in the provided subpopulation.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.ec.vector.FloatVectorSpecies">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.vector.<a href="../../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></h3>
<code><a href="../../../ec/vector/FloatVectorSpecies.html#gaussMutationStdev-int-">gaussMutationStdev</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#inNumericalTypeRange-double-">inNumericalTypeRange</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#loadParametersForGene-ec.EvolutionState-int-ec.util.Parameter-ec.util.Parameter-java.lang.String-">loadParametersForGene</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#maxGene-int-">maxGene</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#minGene-int-">minGene</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationDistributionIndex-int-">mutationDistributionIndex</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationIsBounded-int-">mutationIsBounded</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#mutationType-int-">mutationType</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#outOfRangeRetryLimitReached-ec.EvolutionState-">outOfRangeRetryLimitReached</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#polynomialIsAlternative-int-">polynomialIsAlternative</a>, <a href="../../../ec/vector/FloatVectorSpecies.html#randomWalkProbability-int-">randomWalkProbability</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.ec.vector.VectorSpecies">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.vector.<a href="../../../ec/vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></h3>
<code><a href="../../../ec/vector/VectorSpecies.html#contains-boolean:A-boolean-">contains</a>, <a href="../../../ec/vector/VectorSpecies.html#contains-double:A-double-">contains</a>, <a href="../../../ec/vector/VectorSpecies.html#contains-int:A-int-">contains</a>, <a href="../../../ec/vector/VectorSpecies.html#contains-long:A-long-">contains</a>, <a href="../../../ec/vector/VectorSpecies.html#duplicateRetries-int-">duplicateRetries</a>, <a href="../../../ec/vector/VectorSpecies.html#fill-boolean:A-boolean-">fill</a>, <a href="../../../ec/vector/VectorSpecies.html#fill-double:A-double-">fill</a>, <a href="../../../ec/vector/VectorSpecies.html#fill-int:A-int-">fill</a>, <a href="../../../ec/vector/VectorSpecies.html#fill-long:A-long-">fill</a>, <a href="../../../ec/vector/VectorSpecies.html#initializeGenomeSegmentsByEndIndices-ec.EvolutionState-ec.util.Parameter-ec.util.Parameter-int-">initializeGenomeSegmentsByEndIndices</a>, <a href="../../../ec/vector/VectorSpecies.html#initializeGenomeSegmentsByStartIndices-ec.EvolutionState-ec.util.Parameter-ec.util.Parameter-int-">initializeGenomeSegmentsByStartIndices</a>, <a href="../../../ec/vector/VectorSpecies.html#mutationProbability-int-">mutationProbability</a>, <a href="../../../ec/vector/VectorSpecies.html#setupGenome-ec.EvolutionState-ec.util.Parameter-">setupGenome</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.ec.Species">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.<a href="../../../ec/Species.html" title="class in ec">Species</a></h3>
<code><a href="../../../ec/Species.html#buildMisc-ec.EvolutionState-int-int-">buildMisc</a>, <a href="../../../ec/Species.html#newIndividual-ec.EvolutionState-java.io.DataInput-">newIndividual</a>, <a href="../../../ec/Species.html#newIndividual-ec.EvolutionState-java.io.LineNumberReader-">newIndividual</a>, <a href="../../../ec/Species.html#updateIndividual-ec.EvolutionState-ec.Individual-">updateIndividual</a>, <a href="../../../ec/Species.html#updateSubpopulation-ec.EvolutionState-ec.Subpopulation-">updateSubpopulation</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="P_CMAES_SPECIES">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CMAES_SPECIES</h4>
<pre>public static final&nbsp;java.lang.String P_CMAES_SPECIES</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CMAES_SPECIES">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_LAMBDA">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_LAMBDA</h4>
<pre>public static final&nbsp;java.lang.String P_LAMBDA</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_LAMBDA">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_MU">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_MU</h4>
<pre>public static final&nbsp;java.lang.String P_MU</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_MU">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_SIGMA">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_SIGMA</h4>
<pre>public static final&nbsp;java.lang.String P_SIGMA</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_SIGMA">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_MEAN">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_MEAN</h4>
<pre>public static final&nbsp;java.lang.String P_MEAN</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_MEAN">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_WEIGHTS">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_WEIGHTS</h4>
<pre>public static final&nbsp;java.lang.String P_WEIGHTS</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_WEIGHTS">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_CC">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CC</h4>
<pre>public static final&nbsp;java.lang.String P_CC</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CC">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_CS">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CS</h4>
<pre>public static final&nbsp;java.lang.String P_CS</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CS">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_C1">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_C1</h4>
<pre>public static final&nbsp;java.lang.String P_C1</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_C1">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_CMU">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CMU</h4>
<pre>public static final&nbsp;java.lang.String P_CMU</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_CMU">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_DAMPS">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_DAMPS</h4>
<pre>public static final&nbsp;java.lang.String P_DAMPS</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_DAMPS">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="V_CENTER">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_CENTER</h4>
<pre>public static final&nbsp;java.lang.String V_CENTER</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_CENTER">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="V_ZERO">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_ZERO</h4>
<pre>public static final&nbsp;java.lang.String V_ZERO</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_ZERO">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="V_RANDOM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_RANDOM</h4>
<pre>public static final&nbsp;java.lang.String V_RANDOM</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_RANDOM">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_COVARIANCE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_COVARIANCE</h4>
<pre>public static final&nbsp;java.lang.String P_COVARIANCE</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_COVARIANCE">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="V_IDENTITY">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_IDENTITY</h4>
<pre>public static final&nbsp;java.lang.String V_IDENTITY</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_IDENTITY">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="V_SCALED">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_SCALED</h4>
<pre>public static final&nbsp;java.lang.String V_SCALED</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.V_SCALED">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_ALTERNATIVE_TERMINATION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_ALTERNATIVE_TERMINATION</h4>
<pre>public static final&nbsp;java.lang.String P_ALTERNATIVE_TERMINATION</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_TERMINATION">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_ALTERNATIVE_GENERATOR">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_ALTERNATIVE_GENERATOR</h4>
<pre>public static final&nbsp;java.lang.String P_ALTERNATIVE_GENERATOR</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_GENERATOR">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="P_ALTERNATIVE_GENERATOR_TRIES">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_ALTERNATIVE_GENERATOR_TRIES</h4>
<pre>public static final&nbsp;java.lang.String P_ALTERNATIVE_GENERATOR_TRIES</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.P_ALTERNATIVE_GENERATOR_TRIES">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="lambda">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>lambda</h4>
<pre>public&nbsp;int lambda</pre>
<div class="block">The individuals generated from the distribution. 
        If not specified in the parameters, by default 
        lambda = 4+(int)Math.floor(3*Math.log(n));</div>
</li>
</ul>
<a name="mu">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>mu</h4>
<pre>public&nbsp;int mu</pre>
<div class="block">The truncated individuals used to update the distribution. 
        If not specified in the parameters, by default 
        mu = (int)Math.floor(lambda/2.0);</div>
</li>
</ul>
<a name="weights">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>weights</h4>
<pre>public&nbsp;double[] weights</pre>
<div class="block">The ranked fitness weights for the mu individuals. 
        If not specified in the parameters, by default 
        weights[i] = ln((lambda + 1) / 2i).
        Then all the weights are normalized so that they sum to 1.</div>
</li>
</ul>
<a name="mueff">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>mueff</h4>
<pre>public&nbsp;double mueff</pre>
<div class="block">The "mu_{eff}" constant in CMA-ES.   This is set to 
        1.0 / the sum of the squares of each of the weights[...]</div>
</li>
</ul>
<a name="cmu">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>cmu</h4>
<pre>public&nbsp;double cmu</pre>
<div class="block">The c_{\mu} rank-mu update learning rate.
        If not specified in the parameters, by default
        cmu = Math.min(1-c1, 2 * (mueff - 2 + 1/mueff) / ((n+2)*(n+2) + mueff))
        where n is the genome size.</div>
</li>
</ul>
<a name="c1">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>c1</h4>
<pre>public&nbsp;double c1</pre>
<div class="block">The c_1 rank-1 update learning rate.
        If not specified in the parameters, by default
        c1 = 2 / ((n + 1.3) * (n + 1.3) + mueff)
        where n is the genome size.</div>
</li>
</ul>
<a name="cc">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>cc</h4>
<pre>public&nbsp;double cc</pre>
<div class="block">The c_c rank-one evolution path cumulation parameter.
        If not specified in the parameters, by default
        cc = (4 + mueff / n) / (n + 4 + 2 * mueff/n)
        where n is the genome size.</div>
</li>
</ul>
<a name="cs">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>cs</h4>
<pre>public&nbsp;double cs</pre>
<div class="block">The c_{\sigma} mutation rate evolution path learning rate.
        If not specified in the parameters, by default
        cs = (mueff + 2) / (n + mueff + 5)
        where n is the genome size.</div>
</li>
</ul>
<a name="damps">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>damps</h4>
<pre>public&nbsp;double damps</pre>
<div class="block">The d_{\sigma} dampening for the mutation rate update.
        If not specified in the parameters, by default
        damps = cs + 2 * Math.max(1, Math.sqrt((mueff - 1) / (n + 1))) - 1
        otherwise known as
        damps = 1 + 2 * Math.max(0, Math.sqrt((mueff - 1) / (n + 1)) - 1) + cs
        where n is the genome size.</div>
</li>
</ul>
<a name="sigma">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>sigma</h4>
<pre>public&nbsp;double sigma</pre>
<div class="block">The "sigma" scaling factor for the covariance matrix.</div>
</li>
</ul>
<a name="xmean">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>xmean</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix xmean</pre>
<div class="block">The mean of the distribution.</div>
</li>
</ul>
<a name="c">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>c</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix c</pre>
<div class="block">The "C" covariance matrix of the distribution.</div>
</li>
</ul>
<a name="b">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>b</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix b</pre>
<div class="block">The "B" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</li>
</ul>
<a name="d">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>d</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix d</pre>
<div class="block">The "C" matrix, eigendecomposed from the "C" covariance matrix of the distribution.</div>
</li>
</ul>
<a name="bd">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>bd</h4>
<pre>public&nbsp;org.ejml.data.DenseMatrix64F bd</pre>
<div class="block">b x d</div>
</li>
</ul>
<a name="sbd">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>sbd</h4>
<pre>public&nbsp;org.ejml.data.DenseMatrix64F sbd</pre>
<div class="block">bd x sigma</div>
</li>
</ul>
<a name="invsqrtC">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>invsqrtC</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix invsqrtC</pre>
<div class="block">C^{-1/2}.  This is equal to B x D^{-1} x B^T</div>
</li>
</ul>
<a name="ps">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ps</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix ps</pre>
<div class="block">The p_{\sigma} evolution path vector.</div>
</li>
</ul>
<a name="pc">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>pc</h4>
<pre>public&nbsp;org.ejml.simple.SimpleMatrix pc</pre>
<div class="block">The p_c evolution path vector.</div>
</li>
</ul>
<a name="chiN">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>chiN</h4>
<pre>public&nbsp;double chiN</pre>
<div class="block">An estimate of the expected size of the standard multivariate gaussian N(0,I). 
        This is chiN = Math.sqrt(n)*(1.0-1.0/(4.0*n)+1.0/(21.0*n*n))</div>
</li>
</ul>
<a name="lastEigenDecompositionGeneration">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>lastEigenDecompositionGeneration</h4>
<pre>public&nbsp;int lastEigenDecompositionGeneration</pre>
<div class="block">The most recent generation where an eigendecomposition on C was performed into B and D</div>
</li>
</ul>
<a name="useAltTermination">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>useAltTermination</h4>
<pre>public&nbsp;boolean useAltTermination</pre>
<div class="block">Should we terminate when the eigenvalues get too small?  If we don't, they might go negative and the eigendecomposition will fail.</div>
</li>
</ul>
<a name="useAltGenerator">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>useAltGenerator</h4>
<pre>public&nbsp;boolean useAltGenerator</pre>
<div class="block">If, after trying altGeneratorTries to build an indiviual, we are still building one which violates min/max gene constraints, should
        we instead fill those violated genes with uniformly-selected values between the min and max?</div>
</li>
</ul>
<a name="altGeneratorTries">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>altGeneratorTries</h4>
<pre>public&nbsp;int altGeneratorTries</pre>
<div class="block">How many times should we try to generate a valid individual before we give up and use the useAltGenerator approach?</div>
</li>
</ul>
<a name="DEFAULT_ALT_GENERATOR_TRIES">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>DEFAULT_ALT_GENERATOR_TRIES</h4>
<pre>public static final&nbsp;int DEFAULT_ALT_GENERATOR_TRIES</pre>
<div class="block">Default value (100) for altGeneratorTries.</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.DEFAULT_ALT_GENERATOR_TRIES">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="MAX_TRIES_BEFORE_WARNING">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MAX_TRIES_BEFORE_WARNING</h4>
<pre>public static final&nbsp;int MAX_TRIES_BEFORE_WARNING</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../constant-values.html#ec.eda.cmaes.CMAESSpecies.MAX_TRIES_BEFORE_WARNING">Constant Field Values</a></dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="CMAESSpecies--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>CMAESSpecies</h4>
<pre>public&nbsp;CMAESSpecies()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="defaultBase--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultBase</h4>
<pre>public&nbsp;<a href="../../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;defaultBase()</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../ec/Prototype.html#defaultBase--">Prototype</a></code></span></div>
<div class="block">Returns the default base for this prototype.
        This should generally be implemented by building off of the static base()
        method on the DefaultsForm object for the prototype's package. This should
        be callable during setup(...).</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../ec/Prototype.html#defaultBase--">defaultBase</a></code>&nbsp;in interface&nbsp;<code><a href="../../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code><a href="../../../ec/vector/VectorSpecies.html#defaultBase--">defaultBase</a></code>&nbsp;in class&nbsp;<code><a href="../../../ec/vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></code></dd>
</dl>
</li>
</ul>
<a name="setup-ec.EvolutionState-ec.util.Parameter-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setup</h4>
<pre>public&nbsp;void&nbsp;setup(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                  <a href="../../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from class:&nbsp;<code><a href="../../../ec/Species.html#setup-ec.EvolutionState-ec.util.Parameter-">Species</a></code></span></div>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.  
        If your individual prototype might need to know special things about the species (like parameters stored in it),
        then when you override this setup method, you'll need to set those parameters BEFORE you call super.setup(...),
        because the setup(...) code in Species sets up the prototype.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../ec/Prototype.html#setup-ec.EvolutionState-ec.util.Parameter-">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../ec/Setup.html#setup-ec.EvolutionState-ec.util.Parameter-">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../../ec/Setup.html" title="interface in ec">Setup</a></code></dd>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code><a href="../../../ec/vector/FloatVectorSpecies.html#setup-ec.EvolutionState-ec.util.Parameter-">setup</a></code>&nbsp;in class&nbsp;<code><a href="../../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a></code></dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../ec/Prototype.html#setup-ec.EvolutionState-ec.util.Parameter-"><code>Prototype.setup(EvolutionState,Parameter)</code></a></dd>
</dl>
</li>
</ul>
<a name="clone--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clone</h4>
<pre>public&nbsp;java.lang.Object&nbsp;clone()</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../ec/Prototype.html#clone--">Prototype</a></code></span></div>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.

        <p>Typically this should be a full "deep" clone.
        However, you may share certain elements with other objects
        rather than clone hem, depending on the situation:

        <p>
        <ul>
        <li>If you hold objects which are shared with other instances,
        don't clone them.
        <li>If you hold objects which must be unique, clone them.
        <li>If you hold objects which were given to you as a gesture
        of kindness, and aren't owned by you, you probably shouldn't clone
        them.
        <li> DON'T attempt to clone: Singletons, Cliques, or Populations, or Subpopulation.
        <li>Arrays are not cloned automatically; you may need to
        clone an array if you're not sharing it with other instances.
        Arrays have the nice feature of being copyable by calling clone()
        on them.
        </ul>

        <p><b>Implementations.</b>

        <ul>
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are abstract, then you should not declare clone().
        
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are <b>not</b> abstract, then you should implement
        it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         return super.clone();
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     }
        </pre></tt>
        
        <li>If no ancestor of yours implements clone(), but you
        need to deep-clone some things, then you should implement it
        as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         MyObject myobj = (MyObject) (super.clone());

         // put your deep-cloning code here...
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     return myobj;
     } 
        </pre></tt>

        <li>If an ancestor has implemented clone(), and you also need
        to deep clone some things, then you should implement it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     { 
     MyObject myobj = (MyObject) (super.clone());

     // put your deep-cloning code here...

     return myobj;
     } 
        </pre></tt>
        </ul></div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../ec/Prototype.html#clone--">clone</a></code>&nbsp;in interface&nbsp;<code><a href="../../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code><a href="../../../ec/Species.html#clone--">clone</a></code>&nbsp;in class&nbsp;<code><a href="../../../ec/Species.html" title="class in ec">Species</a></code></dd>
</dl>
</li>
</ul>
<a name="newIndividual-ec.EvolutionState-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>newIndividual</h4>
<pre>public&nbsp;<a href="../../../ec/Individual.html" title="class in ec">Individual</a>&nbsp;newIndividual(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                                int&nbsp;thread)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from class:&nbsp;<code><a href="../../../ec/Species.html#newIndividual-ec.EvolutionState-int-">Species</a></code></span></div>
<div class="block">Provides a brand-new individual to fill in a population.  The default form
        simply calls clone(), creates a fitness, sets evaluated to false, and sets
        the species.  If you need to make a more custom genotype (as is the case
        for GPSpecies, which requires a light rather than deep clone), 
        you will need to override this method as you see fit.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code><a href="../../../ec/vector/VectorSpecies.html#newIndividual-ec.EvolutionState-int-">newIndividual</a></code>&nbsp;in class&nbsp;<code><a href="../../../ec/vector/VectorSpecies.html" title="class in ec.vector">VectorSpecies</a></code></dd>
</dl>
</li>
</ul>
<a name="updateDistribution-ec.EvolutionState-ec.Subpopulation-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>updateDistribution</h4>
<pre>public&nbsp;void&nbsp;updateDistribution(<a href="../../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                               <a href="../../../ec/Subpopulation.html" title="class in ec">Subpopulation</a>&nbsp;subpop)</pre>
<div class="block">Revises the CMA-ES distribution to reflect the current fitness results in the provided subpopulation.</div>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../ec/eda/cmaes/CMAESInitializer.html" title="class in ec.eda.cmaes"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?ec/eda/cmaes/CMAESSpecies.html" target="_top">Frames</a></li>
<li><a href="CMAESSpecies.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
